borsh/src/de/mod.rs

```diff
@@ -472,37 +475,119 @@ pub mod hashes {
     use crate::BorshDeserialize;
     use crate::__maybestd::collections::{HashMap, HashSet};
     use crate::__maybestd::io::{Read, Result};
-    use crate::__maybestd::vec::Vec;
+    use crate::de::hint;
+
+    #[cfg(feature = "de_strict_order")]
+    const ERROR_WRONG_ORDER_OF_KEYS: &str = "keys were not serialized in ascending order";
+    #[cfg(feature = "de_strict_order")]
+    use crate::__maybestd::io::{Error, ErrorKind};
+    #[cfg(feature = "de_strict_order")]
+    use core::cmp::Ordering;
 
     impl<T, H> BorshDeserialize for HashSet<T, H>
     where
-        T: BorshDeserialize + Eq + Hash,
+        T: BorshDeserialize + Eq + Hash + PartialOrd,
         H: BuildHasher + Default,
     {
         #[inline]
         fn deserialize_reader<R: Read>(reader: &mut R) -> Result<Self> {
-            let vec = <Vec<T>>::deserialize_reader(reader)?;
-            Ok(vec.into_iter().collect::<HashSet<T, H>>())
+            let len = u32::deserialize_reader(reader)?;
+
+            if len == 0 {
+                Ok(HashSet::default())
+            } else {
+                let len_hint = hint::cautious::<T>(len);
+                let mut result = HashSet::with_capacity_and_hasher(len_hint, H::default());
+
+                #[cfg(not(feature = "de_strict_order"))]
+                for _ in 0..len {
+                    result.insert(T::deserialize_reader(reader)?);
+                }
+
+                #[cfg(feature = "de_strict_order")]
+                {
+                    let mut prev: Option<T> = None;
+                    for _ in 0..len {
+                        let next_val = T::deserialize_reader(reader)?;
+                        if let Some(ref prev) = prev {
+                            let cmp_result =
+                                prev.partial_cmp(&next_val).map_or(false, Ordering::is_lt);
+                            if !cmp_result {
+                                return Err(Error::new(
+                                    ErrorKind::InvalidData,
+                                    ERROR_WRONG_ORDER_OF_KEYS,
+                                ));
+                            }
+                        }
+
+                        let queued = prev.replace(next_val);
+                        if let Some(queued) = queued {
+                            result.insert(queued);
+                        }
+                    }
+                    if let Some(last_element) = prev {
+                        result.insert(last_element);
+                    }
```

Comment:
Here is a shorter and simpler version (also, potentially, faster):

```suggestion
                    let mut prev = T::deserialize_reader(reader)?;
                    for _ in 1..len {
                        let next_val = T::deserialize_reader(reader)?;
                        let cmp_result = prev.partial_cmp(&next_val).map_or(false, Ordering::is_lt);
                        if !cmp_result {
                            return Err(Error::new(
                                ErrorKind::InvalidData,
                                ERROR_WRONG_ORDER_OF_KEYS,
                            ));
                        }

                        result.insert(prev);
                        prev = next_val;
                    }
                    result.insert(prev);
```

P.S. Please, apply the same to other collections